### PR TITLE
MultiPrecision work, cleaned up

### PR DESCRIPTION
**Pull request by v1kko**

### CODE REVIEW

**Comment by v1kko** [comment]
@fjansson  I think you can have a look at this pull request
- added all the old stuff
- introduced macros, removed the modno files and introduced POIS_PRECISION and FIELD_PRECISION 
- Added intel compilation (with netcdf-fortran from source)

I am only a bit concerned about the macro in modprecision, which is needed to differentiate between gfortran, nag and other compilers (because gfortran and nag run the c preprocessor in "traditional" mode)

File: src/modprecision.f90 [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+! different for them
+#if defined(__GFORTRAN__) || defined(NAGFOR)
+#define PASTE(a) a
+#define CAT(a,b) PASTE(a)b
+#else
+#define PASTE(a,b) a ## b
+#define CAT(a,b) PASTE(a,b)
+#endif
+
+#define SWAP CAT(swap,TYPE_KIND)
```
Comment (v1kko) [inline src/modprecision.f90]: this define is not necessary

File: CMakeLists.txt [hunk 2]
```diff
@@ -154,7 +156,7 @@ FIND_LIBRARY(NETCDF_FORTRAN_LIB netcdff
 )
 
 # libHYPRE.a for static library
-FIND_LIBRARY(HYPRE_LIB HYPRE
+FIND_LIBRARY(HYPRE_LIB HYPRE_core
```
Comment (v1kko) [inline CMakeLists.txt]: re-add the HYPRE target

File: src/modhypre.f90 [hunk 3]
```diff
@@ -19,28 +19,714 @@
 !  Based on hypre/src/test/f77_struct.f
 
 module modhypre
```
Comment (v1kko) [inline src/modhypre.f90:21]: Create an interface to work with real32's currently it just doesn't compile

**Comment by fjansson** [comment]
Testing so far:
on OSX laptop, gfortran 11.2.0, rico-144 benchmark case, 300s, single core.

`export SYST=gnu-fast`

build64/
`cmake .. -DFIELD_PRECISION=64 -DPOIS_PRECISION=64 -DUSE_FFTW=True`

old FFT:
TOTAL wall time =    92.347
1615699968  maximum resident set size
FFTW:
TOTAL wall time =    88.116
1554223104  maximum resident set size

build32/
`cmake .. -DFIELD_PRECISION=32 -DPOIS_PRECISION=32 -DUSE_FFTW=True`

old FFT:
TOTAL wall time =    71.16
1107324928  maximum resident set size
FFTW:
TOTAL wall time =    65.381
1077317632  maximum resident set size

* 32-bit is 23% faster, 31% smaller  (old FFT)
* 32-bit is 26% faster, 31% smaller  (new FFT)
* precision selection flags and FFTW work on gfortran on osx

The ifdef macros look good to me, I will check what the fujitsu compiler makes of them.

**Comment by fjansson** [comment]
Compilation fails on Fugaku with 
`jwd1302i-s  "... dales/src/modprecision.f90", line 17: Not a valid Fortran statement.`
Which is this line: `#if defined(__GFORTRAN__) || defined(NAGFOR)`

Adding `-Cpp` on the Fujitsu compiler command line turns on the preprocessor and solves the problem.

File: src/modprecision.f90 (hunk 1 above)
Comment (v1kko) [reply src/modprecision.f90]: Fixed

File: CMakeLists.txt (hunk 2 above)
Comment (v1kko) [reply CMakeLists.txt]: Fixed

**Pull request merged**